### DISCUSSION TITLE
[Feature request] [Layout animations] Option to skip layout transitions, one-shot layout transitions

### DISCUSSION THREAD

**Discussion by mhoran** [software-mansion/react-native-reanimated#6808] [Ideas]
Similar to #2513, I'm looking for an easier way to selectively disable **layout transitions**. For example, I'd like to animate the layout change of an input box when a button appears in the same row. I achieved this easily with React Native `LayoutAnimation.configureNext` previously. In moving to Reanimated, I've been able to recreate this behavior by reacting to the state change and swapping out the configured `layout` between `LinearTransition` and `undefined`. I've also added a `withCallback` to the animation which disables the animation once it has fired, effectively making it one-shot.

I could see extending `LayoutAnimationConfig` with a `skipTransition` argument. That takes care of half of the problem. Regarding one-shot animations, I have added a single `withCallback` to one of the animations that will fire, which sets the `layout` to `undefined`. I wonder if this should be pulled up higher, e.g. to the `LayoutAnimationConfig` or similar, to help with readability/discoverability.

**Comment by MatiPl01** [comment]
Hey @mhoran!
I raised the topic during our today internal meeting and we decided that this problem is very specific to your use case and we wouldn't want to add a possibility to skip the next layout transition via the `LayoutAnimationConfig`.

If you want to create one-shot layout transitions, you can use your approach with `withCallback` and `useState` and pass `undefined` when you no longer want to use a `LayoutTransition`. You can also write a custom wrapper component if you want to have a simpler and reusable implementation but it depends on your use case.

**Reply by mhoran** [reply]
Thanks @MatiPl01! Appreciate you bringing this forward for discussion.

I had tried to write a custom wrapper component using a `SharedValue` but it didn't work. I don't remember if it was only an issue on New Arch, but disabling the transition *within* the wrapper resulted in the layout getting "stuck". I'll give this another try and open up an issue if I can't get it to work.

**Reply by MatiPl01** [reply]
> Thanks @MatiPl01! Appreciate you bringing this forward for discussion.
> 
> I had tried to write a custom wrapper component using a `SharedValue` but it didn't work. I don't remember if it was only an issue on New Arch, but disabling the transition *within* the wrapper resulted in the layout getting "stuck". I'll give this another try and open up an issue if I can't get it to work. 

You likely won't be able to implement it with SharedValues because layout animations and transitions are tightly coupled with the react's rendering system. I thought that we may be able to implement a possibility to change the layout animation/transition state based on a SharedValue but, after talking about this idea with other teammates, I think that this might be very problematic, so we would prefer to stick to the current implementation in which you can configure layout transitions via component render with a different `layout` prop value.

**Reply by mhoran** [reply]
Hey @MatiPl01, while setting `layout` to `undefined` or the desired transition works as expected on iOS, it doesn't work quite right on Android. Here's a Snack that shows the issue: https://snack.expo.dev/@mhoran/skip-layout-transition-bug (note: Snacks seem pretty broken at the moment -- the simulator is incompatible with the selectable SDK version, which is also quite old. You'll need to run the Snack in some other project.)

On iOS, focusing on the text input causes the layout transition to apply as the button on the right side of the screen appears. The text input smoothly transitions, and then the layout transition is disabled. Typing enough text in the input to cause it to grow to a second line does *not* apply a layout transition.

On Android, focusing on the text input does *not* cause the transition to apply as expected. The text input jumps to the width required by the button on the right side of the screen. Then when typing enough text to cause the text input to grow to a second line, the layout transition applies. Further typing does not apply a layout transition.

There seems to be a difference between iOS and Android when the layout transitions are applied and mutated via props. Any thoughts? Should I open a bug?

**Comment by MatiPl01** [comment]
If you still feel that we should provide an easier way to disable or enable a layout transition, please describe your use case more in detail and we may reconsider adding something in the future.

**Reply by bglgwyng** [reply]
I have implemented a swipable component using reanimated, and I want to animate the exit animation only when the item is deleted, not when it disappears from the active viewport. In this case, I need a way to configure the animation depending on the delete event.

I'm trying this function to acheive it
```ts
function createConditionalLayoutAnimation(
  fn: LayoutAnimationFunction,
  condition: SharedValue<boolean>,
): LayoutAnimationFunction {
  return (values) => {
    "worklet";

    if (!condition.value)
      return {
        animations: {},
        initialValues: {},
      };
    condition.value = false;
    return fn(values);
  };
}
```
It seems to work, but `{ animations: {}, initialValues: {} }` doesn't describe null-animation correctly, so it's a bit buggy yet. Do you think this approach makes sense?

**Reply by bglgwyng** [reply]
I configure the animation setting the value of `condition`.

**Reply by MatiPl01** [reply]
Hey @bglgwyng!

> not when it disappears from the active viewport

What do you mean by saying this? The layout animation is triggered only when the view is unmounted, i.e. is not longer rendered. It is not triggered when the view is still rendered on the screen but invisible (e.g. because of scrolling the scrollable component, etc.)

Btw, mixing SharedValues with layout animations as you do is not a good approach. The layout animation is built while your component renders (if you create it within the component) so you may get an invalid value while reading `condition.value`, because reading SharedValue's value during render (on the JS thread) can give you an old value that is no longer valid.

May you share a code snippet showing the problem you are trying to solve? It would be easier to figure out the solution then.

**Reply by bglgwyng** [reply]
> not when it disappears from the active viewport

I mean, the items in swipable are windowed, so that elements can be removed; not only are their items actually deleted, but they also move out of the window.

Here is my approach https://gist.github.com/bglgwyng/003156c8cd12d00f338ad7a625d07dc5 that configures animation by modifying shared values. The code is not working properly yet.

**Reply by MatiPl01** [reply]
I think that what you are trying to do is to disable the `exiting` animation for views that have already been mounted (rendered for the first time). In general, this is not possible by replacing the `exiting` property by `undefined` as layout animations are configured when the view mounts.

One thing that you may try to use is the `LayoutAnimationConfig`, which makes it possible to enable/disable layout animations for all its children. You can take a look at this example below:

<details>
<summary>Example</summary>

https://github.com/user-attachments/assets/d21d8888-0133-49e5-9ecd-3b9944e11982

```tsx
import React, { useState } from 'react';
import { Button, StyleSheet, View } from 'react-native';
import Animated, {
  FadeOut,
  LayoutAnimationConfig,
} from 'react-native-reanimated';

export default function EmptyExample() {
  const [enabled, setEnabled] = useState(false);
  const [removed, setRemoved] = useState(false);

  return (
    <View style={styles.container}>
      <Button
        title={`${enabled ? 'Disable' : 'Enable'} exiting`}
        onPress={() => setEnabled(!enabled)}
      />
      <Button
        title={removed ? 'Add' : 'Remove'}
        onPress={() => setRemoved(!removed)}
      />
      {!removed && (
        <LayoutAnimationConfig skipExiting={!enabled}>
          <Animated.View style={styles.box} exiting={FadeOut} />
        </LayoutAnimationConfig>
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    gap: 10,
  },
  box: {
    width: 100,
    height: 100,
    backgroundColor: 'red',
  },
});
```
</details>

---

Your code snippet shows that you are using the `layout` prop instead of `exiting` to create the conditional layout transition, so I am a bit confused why you said that you want to animate the exit animation but use the conditional layout transition. Does it somehow impact views that you remove because they are outside of the window bounds?

Btw, you can also enable/disable the layout transition conditionally, like shown in the following code snippet:

<details>
<summary>Example</summary>

https://github.com/user-attachments/assets/fe046c43-681a-478b-a4c0-8fac246dc108

```tsx
import React, { useState } from 'react';
import { Button, StyleSheet, View } from 'react-native';
import Animated, { LinearTransition } from 'react-native-reanimated';

export default function EmptyExample() {
  const [enabled, setEnabled] = useState(false);
  const [width, setWidth] = useState(200);

  return (
    <View style={styles.container}>
      <Button
        title={`${enabled ? 'Disable' : 'Enable'} layout transition`}
        onPress={() => setEnabled(!enabled)}
      />
      <Button
        title={`${width === 100 ? 'Make bigger' : 'Make smaller'}`}
        onPress={() => setWidth(width === 200 ? 100 : 200)}
      />
      <Animated.View
        style={[styles.box, { width }]}
        layout={enabled ? LinearTransition : undefined}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    gap: 10,
  },
  box: {
    width: 100,
    height: 100,
    backgroundColor: 'red',
  },
});
```
</details>

**Reply by MatiPl01** [reply]
I think that in your case, you could just simply wrap the `Reanimated.View` in the map that renders `windowedData` with the `LayoutAnimationConfig` and set the `skipExiting` property to `true` at the same time as you are currently creating this `deleteCond` via `makeMutable(false)`.

**Reply by bglgwyng** [reply]
Updating `skipExiting` requires triggering rerender, so I need to call `flushSync` to ensure the update. I didn't try this, but I think it will work. However, this approach still has issues. One minor issue is the redundant rerendering that makes the performance suboptimal. A more serious issue is that we lack a reliable method to update `skipExiting` before unmounting the element. Sometimes, we need to call `configureDeleteAnimation` thing in the callback function that is called when the windowed data is updated. In that case, we can't update `skipExiting` first and flush, and then update the swipable items later.

**Reply by MatiPl01** [reply]
Ok, I see. Unfortunately, layout animations are tied with react renders and we cannot disable/enable them on the fly without triggering a render. I think that you should stick to tne animated style if possible.

I will also raise the topic during our meeting in the next week and ask if we could make disabling layout animations possible.

**Reply by bglgwyng** [reply]
Thanks!

Also, how do we describe null-exiting-animation? 
Currently, I'm doing it with
```ts
const nullExitingAnimationConfig = {
  animations: {},
  initialValues: {},
};
```
but it didn't work properly. Once I know the correct description of it, then I'll be able to test if my shared value approach works or not.

**Reply by MatiPl01** [reply]
I think I don't understand what you are trying to achieve with this `nullExitingAnimationConfig`. By default, there is no `exiting` animation attached to the view, so you don't have to disable anything. Can you elaborate on what the null-exiting-animation should be?

**Reply by bglgwyng** [reply]
As you mentioned previously, once an element is rendered with certain `exiting` props value, then we should rerender the element to update `exiting` props. I'm trying to avoid it and modifying the animation with a shared value instead. In this case, I need a way to describe null-animation while passing a worklet function to `exiting`.

**Reply by MatiPl01** [reply]
I didn't say that you can update the `exiting` prop when you re-render the component. I said this:

> In general, this is not possible by replacing the exiting property by undefined as layout animations are configured when the view mounts.

I said that you can use the `LayoutAnimationConfig` for this case.

As I mentioned before, there is currently no other way to disable `exiting` animations after your view mounted (rendered for the first time) with an `exiting` animation specified. You cannot use SharedValues for this case as they won't work as expected.

---

I think that also the [example](https://gist.github.com/bglgwyng/003156c8cd12d00f338ad7a625d07dc5) you sent before shows something completely different from what you are trying to achieve. I can see that in this example you create a SharedValue via `makeMutable` in the `windowedData.map` callback, so I wonder how are you trying to avoid re-renders, whilst the `windowedData.map` is executed on every render and `windowedData` changes every time you pass a different `index` via props.

**Reply by bglgwyng** [reply]
Ah, I see. I misunderstood your statement.

> As I mentioned before, there is currently no other way to disable exiting animations after your view is mounted (rendered for the first time) with an exiting animation specified.

Then I think the situation is worse. However, I wasn't trying to achieve disabling layout animation dynamically by updating `exiting` props. `Swipable` example might be quite confusing, as it contains things irrelevant to layout animation.

[Here](https://github.com/bglgwyng/animation-performance-comparison-rn-79/tree/dynamic-animation-example) I made a minimal example of configuring animation using a shared value. The core logic is [here](https://github.com/bglgwyng/animation-performance-comparison-rn-79/blob/dynamic-animation-example/app/(tabs)/index.tsx#L33). You can run the example and see it work.

The reason I was asking how to describe null animation is that disabling animation is just achievable if I can replace `SlideOutRight.build()` in the example with `NullAnimation.build()`. I wanted to implement and show it work, but I couldn't find a way to define null-animation correctly.

Also, I don't think this approach is some kind of hack. I rather think it's the very natural implementation of configuring layout animation. I'd like to hear opinions on it.

**Reply by bglgwyng** [reply]
@MatiPl01 Did you raise this topic in the meeting? Please share any updates if you have.

**Reply by MatiPl01** [reply]
Hey @bglgwyng!

Sorry that I didn't reply. I raised the topic and we decided together that we won't add API to enable/disable layout animations without renders.

I have one note related to what I wrote [here](https://github.com/software-mansion/react-native-reanimated/discussions/6808#discussioncomment-13309663):

> I think that what you are trying to do is to disable the exiting animation for views that have already been mounted (rendered for the first time). In general, this is not possible by replacing the exiting property by undefined as layout animations are configured when the view mounts.

This is a bug and we should be able to remove exiting animations by replacing them with `undefined`. Unfortunately, this still will require an additional render to change this prop.

**Reply by MatiPl01** [reply]
But I also asked if your `SharedValue` approach should work and it should be fine to use a `SharedValue` like this. I also tested if the following code works and everything seems to work fine, at least with latest beta of reanimated 4 with latest RN version.

<details>
<summary>Code snippet</summary>

```tsx
import React, { useMemo, useState } from 'react';
import { Button, StyleSheet, View } from 'react-native';
import type { SharedValue } from 'react-native-reanimated';
import Animated, { useSharedValue, withTiming } from 'react-native-reanimated';

const exitingFactory = (enabled: SharedValue<boolean>) => () => {
  'worklet';
  if (!enabled.value) {
    return {
      animations: {},
      initialValues: {},
    };
  }

  const animations = {
    opacity: withTiming(0, { duration: 1000 }),
  };
  const initialValues = {
    opacity: 1,
  };
  return {
    animations,
    initialValues,
  };
};

export default function EmptyExample() {
  const enabled = useSharedValue(true);
  const [show, setShow] = useState(true);

  const exiting = useMemo(() => exitingFactory(enabled), [enabled]);

  return (
    <View style={styles.container}>
      <Button title="Toggle" onPress={() => setShow((prev) => !prev)} />
      <Button
        title="Toggle enabled"
        onPress={() => (enabled.value = !enabled.value)}
      />
      {show && <Animated.View style={styles.box} exiting={exiting} />}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  box: {
    width: 100,
    height: 100,
    backgroundColor: 'red',
  },
});
```
</details>

Sorry for all the confusion about that but I didn't use such a hacky solution before.

Did you have problems with the exiting animation like this one?

**Reply by bglgwyng** [reply]
Thanks for considering my approach!
I believe this approach isn't a hack but rather an elegant solution.

I also suggest that helper functions like `existingFactory` would be better included in the Reanimated library and its documentation.
Alternatively, if I could get your confirmation that this won't cause any issues, I would feel more confident proceeding with this development direction.

The remaining challenge is how to handle null animations to disable the exiting animation. I haven't figured this out yet. Is it possible to achieve this with the current interface, or would we need to modify the interface?
